File: backend/src/doctor/doctor.entity.ts
import PatientEntity from "src/patient/patient.entity";
import UserEntity from "src/user/user.entity";
import VaccinationEntity from "src/vaccination/vaccination.entity";
import { ChildEntity, Column, Entity, ManyToOne, OneToMany } from "typeorm";
import { JsonProperty, Serializable } from "typescript-json-serializer";
import DoctorPatientConnectionEntity from "./doctor-patient-connection.entity";

@ChildEntity()
@Serializable()
export default class DoctorEntity extends UserEntity {
  @JsonProperty()
  @Column()
  public city: string;

  @JsonProperty()
  @Column()
  public isHomeDoctor: boolean;

  @JsonProperty()
  @Column()
  public isVaccinatorDoctor: boolean;

  @JsonProperty()
  @OneToMany(() => DoctorPatientConnectionEntity, (conn) => conn.doctor, {
    eager: true,
  })
  public doctorPatientConnections: DoctorPatientConnectionEntity[];

  @JsonProperty()
  @OneToMany(() => VaccinationEntity, (vaccination) => vaccination.vaccinator)
  public coordinatedVaccinations: VaccinationEntity[];

  constructor(
    email: string,
    fullName: string,
    phoneNumber: string,
    password: string,
    city: string,
    isHomeDoctor: boolean,
    isVaccinatorDoctor: boolean
  ) {
    super(email, fullName, phoneNumber, password);
    this.city = city;
    this.isHomeDoctor = isHomeDoctor;
    this.isVaccinatorDoctor = isVaccinatorDoctor;
  }
}

File: vaccination_helper/lib/core/doctor/doctor_entity.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:vaccination_helper/core/auth/vaccination/vaccination_entity.dart';
import 'package:vaccination_helper/core/patient/patient_entity.dart';
import 'package:vaccination_helper/core/role/entity/role_entity.dart';
import 'package:vaccination_helper/core/user/user_entity.dart';

part 'doctor_entity.g.dart';

@JsonSerializable()
class DoctorEntity extends UserEntity {
  String city;
  bool isHomeDoctor;
  bool isVaccinatorDoctor;
  List<DoctorPatientConnectionEntity> doctorPatientConnections;
  List<VaccinationEntity> coordinatedVaccinations;

  DoctorEntity();

  static DoctorEntity fromJson(Map<String, dynamic> json) {
    return _$DoctorEntityFromJson(json);
  }

  Map<String, dynamic> toJson() => _$DoctorEntityToJson(this);
}

@JsonSerializable()
class DoctorPatientConnectionEntity {
  String id;
  PatientEntity patient;
  DoctorEntity doctor;
  bool isAccepted;

  DoctorPatientConnectionEntity();

  static DoctorPatientConnectionEntity fromJson(Map<String, dynamic> json) {
    return _$DoctorPatientConnectionEntityFromJson(json);
  }

  Map<String, dynamic> toJson() => _$DoctorPatientConnectionEntityToJson(this);
}
